R/selectors.R

```diff
@@ -101,3 +395,701 @@ cs__by_name <- function(..., .require_all = TRUE) {
     parameters = selector_params
   )
 }
+
+#' Select all categorical columns
+#'
+#' @inherit cs__all return seealso
+#' @examples
+#' df <- pl$DataFrame(
+#'   foo = c("xx", "yy"),
+#'   bar = c(123, 456),
+#'   baz = c(2.0, 5.5),
+#'   .schema_overrides = list(foo = pl$Categorical()),
+#' )
+#'
+#' # Select categorical columns:
+#' df$select(cs$categorical())
+#'
+#' # Select all columns except for those that are categorical:
+#' df$select(!cs$categorical())
+cs__categorical <- function() {
+  wrap_to_selector(pl$col(pl$Categorical()), name = "categorical")
+}
+
+#' Select columns whose names contain the given literal substring(s)
+#'
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Substring(s) that matching
+#' column names should contain.
+#'
+#' @inherit cs__all return seealso
+#' @examples
+#' df <- pl$DataFrame(
+#'   foo = c("x", "y"),
+#'   bar = c(123, 456),
+#'   baz = c(2.0, 5.5),
+#'   zap = c(FALSE, TRUE)
+#' )
+#'
+#' # Select columns that contain the substring "ba":
+#' df$select(cs$contains("ba"))
+#'
+#' # Select columns that contain the substring "ba" or the letter "z":
+#' df$select(cs$contains("ba", "z"))
+#'
+#' # Select all columns except for those that contain the substring "ba":
+#' df$select(!cs$contains("ba"))
+cs__contains <- function(...) {
+  check_dots_unnamed()
+  input <- list2(...)
+  if (!all(vapply(input, is_character, FUN.VALUE = logical(1)))) {
+    abort("All elements of `...` must be of type character.")
+  }
```

Comment:
I believe all inputs are checked when using vapply. Wouldn't it make more sense to use a for loop instead and generate an error the moment there is an unwanted input?

Reply:
I think that would make it more complex for very limited performance gains.

Reply:
I meant: https://stackoverflow.com/questions/23142695/efficient-versions-of-any-all

Reply:
It makes almost no diff for me:
``` r
dat <- as.list(sample(letters, 10000, TRUE))
all_vapply <- function(x) {
  all(vapply(x, is.character, logical(1)))
}
loop <- function(x) {
  for (xi in x) {
    if (!is.character(xi)) {
      return(FALSE)
    }
  }
  TRUE
}
bench::mark(
  all_vapply = all_vapply(dat), 
  loop = loop(dat)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 all_vapply   2.26ms   2.49ms      389.   39.11KB     25.3
#> 2 loop       495.83µs 554.41µs     1742.    3.95MB     26.4

dat[[10001]] <- 1
bench::mark(
  all_vapply = all(vapply(dat, is.character, logical(1))), 
  loop = loop(dat)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 all_vapply   2.23ms   2.51ms      392.    39.1KB     26.3
#> 2 loop       525.54µs  573.3µs     1673.        0B     26.5
```

<sup>Created on 2024-11-12 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
If the loop is terminated in the middle of the process, it will be faster because no extra checks are performed.

``` r
dat <- as.list(sample(letters, 10000, TRUE))
all_vapply <- function(x) {
  all(vapply(x, is.character, logical(1)))
}
loop <- function(x) {
  for (xi in x) {
    if (!is.character(xi)) {
      return(FALSE)
    }
  }
  TRUE
}
bench::mark(
  all_vapply = all_vapply(dat),
  loop = loop(dat)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 all_vapply   4.53ms   4.84ms      194.   39.11KB     13.2
#> 2 loop         1.03ms   1.17ms      727.    3.95MB     11.0

dat[[5000]] <- 1
bench::mark(
  all_vapply = all(vapply(dat, is.character, logical(1))),
  loop = loop(dat)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 all_vapply   4.51ms   4.92ms      185.    39.1KB     13.5
#> 2 loop       512.57µs 667.17µs     1260.        0B     10.6
```

<sup>Created on 2024-11-13 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
That sort of confirms my point, this makes it less readable for a gain < 5ms when 10k patterns are passed, so I don't think this tradeoff is worth it.